sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/PropertyConfigurator.java

```diff
@@ -0,0 +1,144 @@
+
+/**
+ * Allow overriding Part A properties.
+ */
+public class PropertyConfigurator extends AbstractChannelListener {
+
+    private String mAppName;
+
+    private String mAppVersion;
+
+    private String mAppLocale;
+
+    private AnalyticsTransmissionTarget mTransmissionTarget;
+
+    PropertyConfigurator(Channel channel, AnalyticsTransmissionTarget transmissionTarget) {
+        mTransmissionTarget = transmissionTarget;
+        if (channel != null) {
+            channel.addListener(this);
+        }
+    }
+
+    /**
+     *
+     * @param log
+     * @param groupName
+     */
+    @Override
+    public void onPreparingLog(@NonNull Log log, @NonNull String groupName) {
+        if (log instanceof CommonSchemaLog && mTransmissionTarget.isEnabled()) {
+            AppExtension app = ((CommonSchemaLog) log).getExt().getApp();
+
+            /* Override app name if not null, else use the name of the nearest parent. */
+            if (mAppName != null) {
+                app.setId(mAppName);
+            } else {
+                AnalyticsTransmissionTarget parent = mTransmissionTarget.mParentTarget;
+                while (parent != null) {
+                    String parentAppName = parent.getPropertyConfigurator().getAppName();
+                    if (parentAppName != null) {
+                        app.setId(parentAppName);
+                        break;
+                    }
+                    parent = parent.mParentTarget;
+                }
+            }
+
+            /* Override app version if not null, else use the version of the nearest parent. */
+            if (mAppVersion != null) {
+                app.setVer(mAppVersion);
+            } else {
+                AnalyticsTransmissionTarget parent = mTransmissionTarget.mParentTarget;
+                while (parent != null) {
+                    String parentAppVersion = parent.getPropertyConfigurator().getAppVersion();
+                    if (parentAppVersion != null) {
+                        app.setVer(parentAppVersion);
+                        break;
+                    }
+                    parent = parent.mParentTarget;
+                }
+            }
+
+            /* Override app locale if not null, else use the locale of the nearest parent. */
+            if (mAppLocale != null) {
+                app.setLocale(mAppLocale);
+            } else {
+                AnalyticsTransmissionTarget parent = mTransmissionTarget.mParentTarget;
+                while (parent != null) {
+                    String parentAppLocale = parent.getPropertyConfigurator().getAppLocale();
+                    if (parentAppLocale != null) {
+                        app.setLocale(parentAppLocale);
+                        break;
+                    }
+                    parent = parent.mParentTarget;
+                }
+            }
+        }
+    }
+
+    /**
+     * Override common schema Part A property App.Name.
+     *
+     * @param appName   App name
```

Comment:
nit - remove the extra white space

Reply:
dots in parameter names and make sure to commit using Android studio dialog actions, it will decide which spacing to use with the reformat action.

Reply:
This is how auto-reformat formats the code so I'll leave it